### PR TITLE
[SymmMem] Add all_to_all_nd for Ulysses-style exchange

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #178230

Added `torch.distributed._symmetric_memory.all_to_all_nd` with explicit
`scatter_dim` and `gather_dim`:

```
def all_to_all_nd(
    input, out, scatter_dim, gather_dim: int, *, group: str
)
```

`scatter_dim` (int): `0` or `1` — dimension along which the input is partitioned.

`gather_dim` (int): `0` or `1` — dimension along which peer chunks are concatenated in ``out``.

Supported combinations: (1, 0) and (0, 1).

The op avoids a tensor permutation before or after the traditional
all-to-all op (which can only send/receive peer data on dim 0).
It directly reads peers’ shards at strided offsets, from the NCCL
symmetric window and writes a specified output layout, using per-CTA
LSA barriers for synchronization.

Accept 2-D inputs and equivalent row-major 3-D shapes where documented.

Wire the CUDA translation unit into the build and extend distributed
NCCL tests to cover the new API.

@diff-train-skip-merge

### CODE REVIEW

**Comment by codingwithsurya** [comment]
Hi @kwen2501 would love to see this land, would be a great addition to symm mem! Any chance we can get this reviewed?

File: benchmarks/distributed/bench_all_to_all_nd.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            trace_root = os.path.join(os.getcwd(), "bench_a2a_permute_traces")
+            os.makedirs(trace_root, exist_ok=True)
+            trace_dir = tempfile.mkdtemp(prefix="run_", dir=trace_root)
+            trace_json = os.path.join(trace_dir, "col_scatter_both_impls.json")
+            trace_stacks = os.path.join(trace_dir, "col_scatter_both_impls.stacks")
+            prof.export_chrome_trace(trace_json)
+            prof.export_stacks(trace_stacks, metric="self_cuda_time_total")
+
+            table_kwargs = dict(sort_by="self_cuda_time_total", row_limit=30)
+            print()
```
Comment (fduwjj) [inline benchmarks/distributed/bench_all_to_all_nd.py:262]: why do we need this?

File: benchmarks/distributed/bench_all_to_all_nd.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            print(
+                "profile: both per step (all_to_all_permute then permute+all_to_all_single)"
+            )
+            print(prof.key_averages().table(**table_kwargs))
+            print(f"Chrome trace directory: {trace_dir}")
+            print(f"  (chrome://tracing) {trace_json}")
+            print(f"  (export_stacks)    {trace_stacks}")
+            print()
```
Comment (fduwjj) [inline benchmarks/distributed/bench_all_to_all_nd.py:270]: ditto

File: test/distributed/test_nccl.py [hunk 3]
```diff
@@ -579,6 +579,107 @@ def test_reduce_scatter_offset_uneven(self, dim: int):
                 msg=f"rank {self.rank}: out[{j}] should contain the reduced sum",
             )
 
+    @skip_but_pass_in_sandcastle_if(TEST_WITH_ROCM, "Skip NCCL tests for ROCm")
+    @skip_but_pass_in_sandcastle_if(IS_WINDOWS, "NCCL doesn't support Windows")
+    @requires_nccl_version((2, 28, 0), "nccl_all_to_all_permute requires nccl 2.28")
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "scatter_gather,out_2d,input_3d",
```
Comment (fduwjj) [inline test/distributed/test_nccl.py:817]: maybe scatter_gather_dims? But it is too long, so no strong opinion here.

File: torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    int64_t gather_dim,
+    const std::string& group_name) {
+#ifdef NCCL_HAS_SYMMEM_DEVICE_SUPPORT
+  TORCH_CHECK(
+      input.stride(-1) == 1,
+      "nccl_all_to_all_permute: innermost dimension must be contiguous (stride[-1] == 1)");
+  const bool col_scatter =
+      (scatter_dim == 1 && gather_dim == 0);
+  const bool row_scatter =
+      (scatter_dim == 0 && gather_dim == 1);
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu:150]: Do we expect users down the road could have weird tensor shape, so either or both dims can be outside 0 and 1?

File: torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+          "nccl_all_to_all_permute: 3-D input must be row-major contiguous in the last two dimensions "
+          "(stride(1)=local_cols, stride(0)=G*local_cols)");
+      local_cols = static_cast<int>(local_cols_i64);
+    }
+
+    const bool out_shape_3d =
+        out.dim() == 3 && out.size(0) == p && out.size(1) == rows && out.size(2) == local_cols;
+    const bool out_shape_2d = out.dim() == 2 &&
+        out.size(0) == static_cast<int64_t>(p) * static_cast<int64_t>(rows) &&
+        out.size(1) == local_cols;
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu:254]: maybe for my own learning purpose, why do we want to hard code shape_2d/3d here? Is it for perf purpose?

**Review by fduwjj — COMMENTED** [review]
Great PR and I really like Ulysses-style exchange which is way elegant. I think this is also what second order optimizer can use?

**Comment by fduwjj** [comment]
The kernel looks good to me. But my question is that is it possible to make it generic, or we think 2D/3D is good for now?

**Review by fduwjj — APPROVED** [review]
Please make sure torchtitan-CI also pass, unblock for now.

**Comment by amdfaa** [comment]
**Heads up: please rebase to fix CI runner scheduling**

CI jobs on this PR may be failing to schedule because they're requesting runner labels that don't exist - e.g. `lf.amd-do.linux.c7i.2xlarge` (the `amd-do` experiment prefix is leaking onto non-ROCm runners).

The runner-determinator script is checked out from this PR's head commit, not from `main`. Because this branch predates #186141, your jobs run the old version of `runner_determinator.py`, which incorrectly folds the `amd-do` experiment into the shared runner-label prefix for every job instead of scoping it to ROCm runners.

Please rebase onto the latest `main` (which includes #186141) and re-push. Your jobs will then pick up the fixed runner-determinator and schedule on the correct runners.

**Comment by ngimel** [comment]
Late to the party, but can we make it `all_to_all_nd` instead of `permute`? There's no permutation taking place.

**Comment by malfet** [comment]
> ## Reverting PR 178230 failed
> **Reason**: No revertable PRs found in ghstack for #178230. This typically means the PR is still open (not merged) or its GitHub state is inconsistent. Only closed/merged PRs can be reverted.
> 
> Details for Dev Infra team

Not it did not, see https://github.com/pytorch/pytorch/commit/222c5cad5bda3da2965e1e7148ee3b840a8e2678

**Comment by kwen2501** [comment]
@malfet @ngimel Sorry this PR broke trunk. 
Previously this PR's lint is green, so I did a "-f" after a simple rebase. Bad idea. 

What's weird tho: the PR didn't touch in the file in question: 
`>>> Lint for torch/csrc/distributed/c10d/TraceUtils.h`.
which is confirmed by Claude's analysis above. 

In any case, if need be, I will address the lint in my PR.

**Comment by kwen2501** [comment]
> can we make it all_to_all_nd instead of permute

@ngimel Makes sense. Renaming. Thanks!

**Comment by ngimel** [comment]
You need to rebase to main, linter was previously broken, so now unrebased PRs are stopped by linter

**Comment by kwen2501** [comment]
@malfet This PR fixes the Lint error upstream: https://github.com/pytorch/pytorch/pull/187706
Can you please take a look? 
I realized that it would be cleaner to fix it in a separate PR than in this one. Thanks!

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but it seems to fail a bunch of internal NCCL tests' -c ghfirst

There is one from ExecuTorch for example

```
buck build --flagfile fbcode//mode/dev fbcode//executorch/examples/models/fb/gemma4:run_gemma4
```

failed with

```
Buck build failed for this target, and is likely caused by your changes.
Error message:
[New Failure] Target failed to build.
Action failed: fbcode//caffe2:_libtorch_cuda (cfg:core_build_mode-linux-x86_64-fbcode-platform010-clang19-no-san-opt-by-default#a39db326b839c336) (cudafepp torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu (pic))
Remote command returned non-zero exit code 2
Remote action, reproduce with: `frecli cas download-action df336451bff0483ad16082e531321fb540bfbd302e02d7c787857a1f894c9ca5:148`
Stdout: <empty>
Stderr:
fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(99): error: identifier "ncclCoopCta" is undefined
    const ncclCoopCta coop{};
          ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(101): error: identifier "ncclCoopCta" is undefined
    ncclLsaBarrierSession<ncclCoopCta> bar{
                          ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(104): error: identifier "ncclTeamLsa" is undefined
        ncclTeamLsa(devComm),
        ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(105): error: class "ncclDevComm" has no member "lsaBarrier"
        devComm.lsaBarrier,
                ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(107): error: namespace "cuda" has no member "memory_order_acquire"
    bar.sync(coop, cuda::memory_order_acquire);
                         ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(118): error: identifier "ncclGetLsaPointer" is undefined
      const void* src_row = ncclGetLsaPointer(window, row_byte_offset, peer_idx);
                            ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(130): error: namespace "cuda" has no member "memory_order_release"
    bar.sync(coop, cuda::memory_order_release);
                         ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(175): error: identifier "ncclDevCommRequirements" is undefined
      ncclDevCommRequirements reqs = NCCL_DEV_COMM_REQUIREMENTS_INITIALIZER;
      ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(175): error: identifier "NCCL_DEV_COMM_REQUIREMENTS_INITIALIZER" is undefined
      ncclDevCommRequirements reqs = NCCL_DEV_COMM_REQUIREMENTS_INITIALIZER;
                                     ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(178): error: identifier "ncclDevCommCreate" is undefined
      do { ncclResult_t result = ncclDevCommCreate(comm, &reqs, &devcomm); if (result != ncclSuccess) { std::string err = "NCCL error in: " + std::string("fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu") + ":" + std::to_string(180) + ", " + ncclGetErrorWithVersion(result) + "\n" + getNcclErrorDetailStr(result, "ncclDevCommCreate failed in nccl_all_to_all_nd"); if (!(false)) { throw ::c10::DistBackendError( {__func__, "fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu", static_cast<uint32_t>(180)}, (::c10::detail::torchCheckMsgImpl( "Expected " "false" " to be true, but got false.  " "(Could this error message be improved?  If so, " "please report an enhancement request to PyTorch.)", err))); }; } } while (0);
                                 ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(185): error: class "ncclDevComm" has no member "rank"
    const int my_rank = devcomm.rank;
                                ^

fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu(186): error: class "ncclDevComm" has no member "nRanks"
    const int p = devcomm.nRanks;
                          ^

12 errors detected in the compilation of "fbcode/caffe2/torch/csrc/distributed/c10d/symm_mem/ops/nccl_all_to_all_nd.cu".
```

It seems like internal uses an older NCCL. Maybe someone from distributed team can chime in

**Comment by kwen2501** [comment]
Short version:

The device-side symbols this kernel uses (`ncclCoopCta`, `ncclTeamLsa`, `ncclDevComm::lsaBarrier`, `ncclGetLsaPointer`, ...) are present in public NCCL 2.28.x -- I checked the headers and compiled against them.

So the internal failure looks like a toolchain that's effectively **older than the version number it reports**: `NCCL_VERSION_CODE` says `>= 2.28.4` (so our `#ifdef` lets the code compile in), but the internal headers don't actually define those symbols, and `cuda::memory_order` is missing too (older libcu++/CUDA). The version number says "supported" while the headers say "not here."

A version check can't catch that -- the number is lying. The robust fix is to stop trusting the version macro and **probe at build time**: try to compile a tiny snippet that uses the API, and enable the op only if it compiles. On a toolchain that lacks it, the op then compiles out cleanly instead of breaking the build.

@huydhn @georgehong -- is internal pinned to an NCCL/CUDA older than the version it advertises (or NCCL built with the device API off)? If so, a build-time probe would stop this from breaking internal builds.

**Comment by huydhn** [comment]
@huydhn has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D109369584).

**Comment by huydhn** [comment]
@georgehong I have imported this change in D109369584.  Let's wait a bit for the signals.  We can ask @kapilsh and @d4l3k to help if the NCCL failures show up

**Comment by kwen2501** [comment]
@huydhn @georgehong 
I am bumping the gate from 2.28 to 2.29, for a separate reason -- a host-side initializer requires 2.29.
And I notice that another op `reduce_scatter_offset` gated at 2.29 did not trigger Meta's internal error.
So I guess with 2.29 I should be fine. With that I am trying to reland.

**Comment by huydhn** [comment]
I will need to reimport this again with the latest change as the content need to match

**Comment by huydhn** [comment]
@huydhn has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D109369584).

**Comment by huydhn** [comment]
@huydhn has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D109369584).

**Comment by huydhn** [comment]
Ok, I'm landing the diff then
